### PR TITLE
Add impression(component:view) tracking

### PR DESCRIPTION
**Pull request by fenglish**
Component view events will be fired with elements set `data-o-tracking-view` attribute or elements collected by a custom selector in `opts`. They will populate a `domPathTokens` as well as Click events. You could pass a custom category, and also a callback[`getContextData`] to collect data for events.  

:warning: This feature requires `window.IntersectionObserver` in order to track the events  
:warning: `getContextData` should be a function which returns `{Object}`

```js
const opts = {
	category: 'audio', // default: 'component'
	selector: '.o-teaser__audio', // default: '[data-o-tracking-view]'
	getContextData: () => {  // default: null
		return {
			contentId: '1234',
			type: 'audio',
			subtype: 'podcast',
		 	component: 'teaser'
		};
	}
}

oTracking.view.init(opts);
```

### CODE REVIEW

File: src/javascript/events/component-view.js [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+ * @alias view#init
+ * @param {Object} opts - To set custom category, elementsToTrack, callback to collect data for the event
+ * @return {undefined}
+ */
+const init = (opts = {}) => {
+	if(!window.IntersectionObserver) {
+		return false;
+	}
+
+	const elementsToTrack = opts.elementsToTrack || [...document.querySelectorAll('[data-o-tracking-view]')];
```
Comment (robsquires) [inline src/javascript/events/component-view.js]: I'm wondering if passing a selector here might be simpler for the applications using this + possibly less prone to misconfiguration (e.g. what if you don't pass an array of HTMLElements) ? 🤔

File: README.md [hunk 2]
```diff
@@ -188,6 +192,16 @@ o-tracking will
 * Automatically pickup ftsession from cookies for you.
 * Page events automatically pick up the url and the referrer.
 * Click events [initalised as above](#usage), will populate a `domPathTokens` property. If the clicked element has the `data-trackable` attribute, sibling elements will also be included within `domPathTokens`.
+* Component view events will be fired with elements set `data-o-tracking-view` attribute or passed as elementsToTrack[array] in `opts`. They will populate a `domPathTokens` as well as Click events. You could pass custom category and also callback to collect data for events.  
+	```js
+	const opts = {
+		category: 'custom', // default: 'component'
+		elementsToTrack: [ element, element, ... ], // default: elements set data-o-tracking-view
+		callback: () => {} // default: null
```
Comment (robsquires) [inline README.md]: Wondering if we could make the name of this more explicit, something like ... `getContextForComponent`  🤔 ?

File: test/events/component-view.test.js [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+			targetComponent = document.createElement('div');
+			targetComponent.setAttribute('data-o-tracking-view', true);
+			targetComponent.innerHTML = 'Default component view elements';
+			document.body.appendChild(targetComponent);
+
+			componentView.init();
+			viewed(targetComponent);
+		});
+
+		it('should track an event for a component view ', () => {
```
Comment (robsquires) [inline test/events/component-view.test.js]: Would it be worth adding a case for `should only track an event for a component view once` ?

File: test/events/component-view.test.js [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+			targetComponent.innerHTML = 'Custom component view elements';
+			targetComponent.setAttribute(targetAttribute, true);
+			targetComponent.setAttribute(idAttribute, id);
+			document.body.appendChild(targetComponent);
+
+			const opts = {
+				category,
+				elementsToTrack: [...document.querySelectorAll(`[${targetAttribute}]`)],
+				callback: (el) => {
+					optsCallbackSpy(el);
```
Comment (robsquires) [inline test/events/component-view.test.js]: Minor thing, but given you're asserting on the data returned by this function, you could probably eschew spying on this function as well.

File: src/javascript/events/component-view.js [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+	function onChange (changes) {
+		changes.forEach(change => {
+			if(change.isIntersecting || change.intersectionRatio > 0) {
+				const eventData = {
+					action: 'view',
```
Comment (benbarnett) [inline src/javascript/events/component-view.js]: I know we sort of settled on this now but I'm wondering if `component:seen` might be clearer........ what do you think? No worries if we're too far down the line!

Comment (fenglish) [reply src/javascript/events/component-view.js]: @benbarnett The business analyst side wants to keep consistent for the action.
`view` is already used for impressions. e.g. comment:view.

File: test/events/component-view.test.js (hunk 3 above)
Comment (fenglish) [reply test/events/component-view.test.js]: @robsquires I thought That is a IntersectionObserver side test. It might be a bit tricky to test that?
It means `mockIntersectionObserver` has to have the code for `unobserve`.

There is a line to check `unobserve` is called. Is it not enough?
`assert.equal(unobserveSpy.calledOnce, true, "IntersectionObserver unobserved target");`

Comment (robsquires) [reply test/events/component-view.test.js]: Could you fire `viewed(targetComponent);` twice?

File: src/javascript/events/component-view.js [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+const Core = require('../core');
+const getTrace = require('../../libs/get-trace');
+const utils = require('../utils');
+
+const decorateEventData = (eventData, viewedEl, opts) => {
+
+	const { trace, customContext } = getTrace(viewedEl);
+	let context = typeof opts.getContextData === 'function' ? opts.getContextData(viewedEl) : {};
+
+	if (typeof context !== 'object') {
```
Comment (benbarnett) [inline src/javascript/events/component-view.js]: Is this block needed?

I guess this is guarding against consumers providing an invalid callback function, i.e. one that doesn't return an object. If that is the case then I think this could probably make people think they are using this method correctly, when in fact it should be erroring...

File: src/javascript/utils.js [hunk 7]
```diff
@@ -214,6 +214,30 @@ function getValueFromJsVariable(str) {
 	return test !== '' ? test : null;
 }
 
+/**
+ * Trim strings
+ * @param {String} str - The string to trim.
+ * @return {String} The trimed string.
```
Comment (benbarnett) [inline src/javascript/utils.js]: typo: `trimmed`

File: src/javascript/utils.js [hunk 8]
```diff
@@ -233,5 +257,7 @@ module.exports = {
 	triggerPage: triggerPage,
 	getValueFromCookie: getValueFromCookie,
 	getValueFromUrl: getValueFromUrl,
-	getValueFromJsVariable: getValueFromJsVariable
+	getValueFromJsVariable: getValueFromJsVariable,
+	sanitise: sanitise,
+	assignIfUndefined: assignIfUndefined
```
Comment (benbarnett) [inline src/javascript/utils.js]: Looks like we have a mix of trailing commas and no trailing commas (i.e. on `src/libs/get-trace.js`)

If it this is true for the rest of the repo perhaps it could be cleaned up in a separate PR.

File: test/events/component-view.test.js [hunk 9]
```diff
@@ -0,0 +1,208 @@
+/*global require, describe, it, before, after, sinon, document */
+require('../setup');
+const assert = require("assert");
+const settings = require("../../src/javascript/core/settings");
```
Comment (benbarnett) [inline test/events/component-view.test.js]: These imports typically use single quotes

File: src/javascript/events/component-view.js [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+
+/**
+ * Listen for view events.
+ *
+ * @alias view#init
+ * @param {Object} opts - To set custom category[String], selector[String], getContextData[Function]
+ * @return {undefined}
+ */
+const init = (opts = {}) => {
+	if(!window.IntersectionObserver) {
```
Comment (benbarnett) [inline src/javascript/events/component-view.js]: Is it worth adding somewhere to the readme that this feature requires `window.IntersectionObserver` in order to track the events? https://caniuse.com/#feat=intersectionobserver

In FT.com we are loading the polyfill in n-ui, so not a problem 😀

File: src/javascript/events/component-view.js [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+'use strict'; // eslint-disable-line strict
+
+const Core = require('../core');
+const getTrace = require('../../libs/get-trace');
+const utils = require('../utils');
+
+const decorateEventData = (eventData, viewedEl, opts) => {
+
+	const { trace, customContext } = getTrace(viewedEl);
+	let context = typeof opts.getContextData === 'function' ? opts.getContextData(viewedEl) : {};
```
Comment (robsquires) [inline src/javascript/events/component-view.js]: I'm wondering if it's worth running the object returned from `getContextData` through a whitelist like [o-audio](https://github.com/Financial-Times/o-audio/blob/master/src/js/tracking.js#L46). In practice these attributes need to be agreed with the data team + would provide some free documentation as to what can be returned. What ya think?

File: src/javascript/utils.js (hunk 7 above)
Comment (fenglish) [reply src/javascript/utils.js]: 🙈

File: test/events/component-view.test.js (hunk 9 above)
Comment (fenglish) [reply test/events/component-view.test.js]: Oh thank you!

File: src/javascript/events/component-view.js (hunk 10 above)
Comment (fenglish) [reply src/javascript/events/component-view.js]: Oh,That is a good point. I'm going to add the explanation.

File: src/javascript/events/component-view.js (hunk 6 above)
Comment (fenglish) [reply src/javascript/events/component-view.js]: I see. I will keep the block because if the function returns String, it won't throw an error, just will mess up the data detail. So I will use the block to throw an error.

File: src/javascript/events/component-view.js (hunk 11 above)
Comment (fenglish) [reply src/javascript/events/component-view.js]: I see! I will do that. Thank you!

File: src/javascript/events/component-view.js (hunk 6 above)
Comment (benbarnett) [reply src/javascript/events/component-view.js]: Sounds sensible!

File: test/events/component-view.test.js [hunk 12]
```diff
@@ -0,0 +1,226 @@
...
+	context: {
+		product: 'desktop'
+	},
+	user: {
+		user_id: '123456'
+	},
+};
+
+let targetComponent;
+let error;
```
Comment (benbarnett) [inline test/events/component-view.test.js]: I think this is good as it is, but do you think this would be marginally clearer if it was called `errorThrown`?

File: test/utils.test.js [hunk 13]
```diff
@@ -70,4 +70,42 @@ describe('Utils', function () {
...
+			'contentId',
+			'subtype',
+			'type',
+		];
+		const resultProps = { contentId: 1234, subtype: 'subtype' };
+		const testProps = Object.assign(
+			{},
+			resultProps,
+			{ type: undefined, name: 'name' }
+		);
```
Comment (benbarnett) [inline test/utils.test.js]: Tiny thing again but you could refactor this to match the array style in the rest of this spec?

Comment (fenglish) [reply test/utils.test.js]: 👍

File: test/events/component-view.test.js (hunk 12 above)
Comment (fenglish) [reply test/events/component-view.test.js]: I agree!

File: src/javascript/events/component-view.js (hunk 10 above)
Comment (JakeChampion) [reply src/javascript/events/component-view.js]: It should be added to the `origami.json.browserFeatures.required` array. --> https://origami.ft.com/docs/syntax/origamijson/

Comment (fenglish) [reply src/javascript/events/component-view.js]: @JakeChampion Thank you for telling me!

File: README.md [hunk 14]
```diff
@@ -188,6 +192,25 @@ o-tracking will
 * Automatically pickup ftsession from cookies for you.
 * Page events automatically pick up the url and the referrer.
 * Click events [initalised as above](#usage), will populate a `domPathTokens` property. If the clicked element has the `data-trackable` attribute, sibling elements will also be included within `domPathTokens`.
+* Component view events will be fired with elements set `data-o-tracking-view` attribute or elements collected by a custom selector in `opts`. They will populate a `domPathTokens` as well as Click events. You could pass a custom category, and also a callback[`getContextData`] to collect data for events.  
+:warning: This feature requires `window.IntersectionObserver` in order to track the events
```
Comment (notlee) [inline README.md]:
```suggestion
_Note:_ This feature requires `window.IntersectionObserver` in order to track the events.
```

This is super trivial, but maybe we could say "note" rather than "warning" here, and below for `getContextData`? Warnings are scary 🙈

File: README.md [hunk 15]
```diff
@@ -188,6 +192,25 @@ o-tracking will
...
 * Page events automatically pick up the url and the referrer.
 * Click events [initalised as above](#usage), will populate a `domPathTokens` property. If the clicked element has the `data-trackable` attribute, sibling elements will also be included within `domPathTokens`.
+* Component view events will be fired with elements set `data-o-tracking-view` attribute or elements collected by a custom selector in `opts`. They will populate a `domPathTokens` as well as Click events. You could pass a custom category, and also a callback[`getContextData`] to collect data for events.  
+:warning: This feature requires `window.IntersectionObserver` in order to track the events  
+:warning: `getContextData` should be a function which returns `{Object}`
+	```js
+	const opts = {
+		category: 'audio', // default: 'component'
+		selector: '.o-teaser__audio', // default: '[data-o-tracking-view]'
+		getContextData: () => {  // default: null
```
Comment (notlee) [inline README.md]: If I understand correctly, the callback function accepts the viewed element as an argument? It would be great to document that here.

File: src/javascript/events/component-view.js [hunk 16]
```diff
@@ -0,0 +1,86 @@
...
+
+/**
+ * Listen for view events.
+ *
+ * @alias view#init
+ * @param {Object} opts - To set custom category[String], selector[String], getContextData[Function]
+ * @return {undefined}
+ */
+const init = (opts = {}) => {
+	if(!window.IntersectionObserver) {
```
Comment (notlee) [inline src/javascript/events/component-view.js]: It's good that this is optional to avoid a breaking change :) But perhaps a warning here would be good to help debug, if users expect to track views across older browsers?
```
console.warn('o-tracking: Unable to track element view events as "window.IntersectionObserver" is not supported.');
```

Comment (fenglish) [reply src/javascript/events/component-view.js]: 👍

File: README.md (hunk 15 above)
Comment (fenglish) [reply README.md]: 👍

File: README.md (hunk 14 above)
Comment (fenglish) [reply README.md]: 🙈

File: README.md [hunk 17]
```diff
@@ -188,6 +192,25 @@ o-tracking will
 * Automatically pickup ftsession from cookies for you.
 * Page events automatically pick up the url and the referrer.
 * Click events [initalised as above](#usage), will populate a `domPathTokens` property. If the clicked element has the `data-trackable` attribute, sibling elements will also be included within `domPathTokens`.
+* Component view events will be fired with elements set `data-o-tracking-view` attribute or elements collected by a custom selector in `opts`. They will populate a `domPathTokens` as well as Click events. You could pass a custom category, and also a callback[`getContextData`] to collect data for events.
```
[the remark this answers is not in the document]
Comment (fenglish) [reply README.md]: I see. Thank you for making them nicer!
